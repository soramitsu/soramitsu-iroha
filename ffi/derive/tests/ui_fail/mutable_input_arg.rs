//! Test which assures that methods taking mutable references as input arguments are not
//! allowed in the FFI. This limitation can be lifted in the future, and consequently
//! this test nullified, if the need for mutable input arguments feature arises

use iroha_ffi::{ffi_bindgen, FfiResult};
use std::mem::MaybeUninit;

struct FfiStruct {
    a: u32,
}

#[ffi_bindgen]
impl FfiStruct {
    /// From mutable input arg
    pub fn from_mutable_input_arg(a: &mut u32) -> Self {
        let output = Self { a: a.clone() };
        *a = 42;
        output
    }
}

fn main() -> Result<(), ()> {
    let s: MaybeUninit<*mut FfiStruct> = MaybeUninit::uninit();

    if FfiResult::Ok != FfiStruct__from_mutable_input_arg(s.as_mut_ptr()) {
        return Err(());
    }

    Ok(())
}
