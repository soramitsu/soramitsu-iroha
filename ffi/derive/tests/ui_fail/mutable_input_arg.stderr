error: Mutable references not supported
  --> tests/ui_fail/mutable_input_arg.rs:15:38
   |
15 |     pub fn from_mutable_input_arg(a: &mut u32) -> Self {
   |                                      ^^^^^^^^

error[E0425]: cannot find function, tuple struct or tuple variant `FfiStruct__from_mutable_input_arg` in this scope
  --> tests/ui_fail/mutable_input_arg.rs:25:25
   |
25 |     if FfiResult::Ok != FfiStruct__from_mutable_input_arg(s.as_mut_ptr()) {
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
